addons/maintenance/models/maintenance.py

```diff
@@ -190,17 +193,62 @@ def _read_group_category_ids(self, categories, domain, order):
         category_ids = categories._search([], order=order, access_rights_uid=SUPERUSER_ID)
         return categories.browse(category_ids)
 
+    def _create_new_request(self, date):
+        self.ensure_one()
+        self.env['maintenance.request'].create({
+                    'name': _('Preventive Maintenance - %s') % self.name,
+                    'request_date': fields.Date.to_string(date),
+                    'schedule_date': fields.Date.to_string(date),
+                    'category_id': self.category_id.id,
+                    'equipment_id': self.id,
+                    'maintenance_type': 'preventive',
+                    'owner_user_id': self.owner_user_id.id,
+                    'technician_user_id': self.technician_user_id.id,
+                    })
+
+    def _fill_up(self, from_date, to_date, to_end=False):
+        """
+            Will add maintenance requests between from_date and to_date with the equipment maintenance period as interval
+            Does nothing when the interval is too small
+            
+            to_end determines whether you need to create one at the end also or whether at the 
+            end there is a maintenance request already
+        """
+        self.ensure_one()
+        period = self.period
+        current_date = from_date + relativedelta(days=period)
+        while (current_date <= to_date):
+            if to_end or current_date + relativedelta(days=period) <= to_date:
+                self._create_new_request(current_date)
+            current_date = current_date + relativedelta(days=period)
+
     @api.model
     def _cron_generate_requests(self):
-        for equipment in self.search([]):
-            if equipment.period and equipment.next_action_date == date.today().strftime(DEFAULT_SERVER_DATE_FORMAT):
-                self.env['maintenance.request'].create({
-                    'name': _('Preventive Maintenance - %s' % equipment.next_action_date),
-                    'request_date': equipment.next_action_date,
-                    'category_id': equipment.category_id.id,
-                    'equipment_id': equipment.id,
-                    'maintenance_type': 'preventive',
-                })
+        show_until = fields.Date.from_string(fields.Datetime.now()) + relativedelta(months=1)
```

Comment:
I have some concerns about this algorithm. Code is complicated and I think it can be improved

 - there is a mismatch in dates management: you switch date to the future or the past by period (once in _cron_generate, once in _fill_up). I think you can remove those switches and simplify understanding
 - remove the _to_end flag that make the code difficult to understand; sometimes you include a last interval, sometimes not. By looking at the code you may have cases where time between two requests is greater than period. I think the purpose of preventive maintenance is to ensure request delays are <= period, not > .
 - in _fill_up, use rrule to ease understanding and rely on standard library

Keeping this in mind I think you could reduce the code and simplify the algorithm. Thanks!